File: Content/Mods/StyxScribeShared/StyxScribeShared.py
from weakref import WeakKeyDictionary
from numbers import Number
from collections import OrderedDict
from inspect import getmembers
from functools import wraps
from threading import local as thread_local
from types import MethodType

proxyTypes = dict()
marshallTypes = OrderedDict()
DELIM = '¦'

registry = None
lookup = None

NIL = None
Root = None

#https://stackoverflow.com/a/49013535
_meta_ignore = {"__repr__","__format__","__str__","__reduce__","__reduce_ex__","__sizeof__"}
_meta_inherited = {}
def _meta_wrap(name, method):
    local_flag = thread_local()
    @wraps(method)
    def wrapper(*args, **kw):
        local_method = method
        if not getattr(local_flag, "running", False) and args and not isinstance(args[0], type):
            local_flag.running = True
            # trigger __getattribute__:
            self = args[0]
            cls = self.__class__
            retrieved = cls.__getattribute__(self, name)
            if not retrieved is wrapper:
                local_method =  retrieved
            if isinstance(local_method, MethodType):
                args = args[1:]
        result = local_method(*args, **kw)
        local_flag.running = False
        return result
    wrapper._wrapped = True
    return wrapper
class MetaOverrider(type):
    def __init__(cls, name, bases, namespace, **kwd):
        super().__init__(name, bases, namespace, **kwd)
        
        global _meta_inherited
        d = None
        for c in cls.__mro__:
            if "Proxy" not in map(lambda _c: _c.__name__, c.__mro__):
                _cls = c
                d = inherited(_cls, cls)
                break
        _meta_inherited[cls] = d

        for name in dir(cls):
            if not (name.startswith("__") and name.endswith("__")):
                continue
            if name in ("__getattribute__", "__class__", "__init__"):
                continue
            if name in object.__dict__ and name not in _meta_ignore:
                continue
            if d is not None and name not in d:
                continue
            magic_method = getattr(cls, name)
            if not callable(magic_method) or getattr(magic_method, "_wrapped", False):
                continue
            setattr(cls, name, _meta_wrap(name, magic_method))

def _address(obj):
    s = ['0'] * 16
    h = hex(id(obj))[2:].upper()
    s[-len(h):] = h
    return "0x" + ''.join(s)
    
def _repr(obj, name=None):
    mid = f" {name}" if name is not None else ""
    return f"<{obj.__class__.__qualname__}{mid} at {_address(obj)}>"

def nop():
    pass

def inherited(a,b):
    #https://stackoverflow.com/a/48281090
    a = dict(getmembers(a))
    b = dict(getmembers(b))
    return {k:v for k,v in b.items() if k in a and v == a[k]}

class _function():
    def __init__(self,call=None):
        if call is None:
            call = nop
        self.call = call
    def __call__(self,*args,**kwargs):
        return self.call(*args,**kwargs)
    def __repr__(self):
        return repr(self.call)

def notImplemented(self, *args, **kwargs):
    raise NotImplementedError(f"{type(self)} does not implement this operation")

def proxyType(cls):
    proxyTypes[cls.__name__] = cls
    return cls

def marshallType(*types):
    def f(cls):
        for t in types:
            mt = marshallTypes.get(cls,set())
            mt.add(t)
            marshallTypes[cls] = mt
        return cls
    return f

class Proxy(metaclass=MetaOverrider):
    def __init__(self, v=None, i=None):
        #initialise a new object to proxy, select from inheritance
        for cls in type(self).__mro__:
            if Proxy not in cls.__mro__:
                proxy = cls()
                break
        if i is None:
            i = id(self)
        self._proxy = proxy
        registry[i] = self
        lookup[self] = i
        self._root = i == 0
        self._local = i > 0
        self._alive = True
        if self._local:
            Scribe.Send(f"StyxScribeShared: New: {self.__class__.__name__}{DELIM}{i}")
        if v is not None:
            self._marshall(v)
    def __hash__(self):
        return hash(id(self))
    def __repr__(self):
        if hasattr(self,"_name"):
            return _repr(self, self._name)
        return _repr(self)
    def __del__(self):
        if lookup is None:
            return
        i = lookup.get(self,None)
        if i is not None:
            self._alive = False
            try:
                del lookup[self]
            except KeyError:
                pass
            if i != 0:
                try:
                    del registry[i]
                except KeyError:
                    pass
                if self._local:
                    Scribe.Send(f"StyxScribeShared: Del: {i}")
    def __getattribute__(self, name):
        if name in _meta_inherited[type(self)]:
            proxy = object.__getattribute__(self, "_proxy")
            try:
                attr = getattr(proxy, name)
                if callable(attr) and attr.__self__ == proxy:
                    def attr_wrapper(*args,**kwargs):
                        return attr(*args[1:],**kwargs)
                    return attr_wrapper
                return attr
            except AttributeError:
                pass
        return object.__getattribute__(self, name)       

class ProxySet(Proxy):
    _repr = False
    def _shset(self, key, val):
        if self._alive:
            i = lookup[self]
            k = encode(key)
            v = encode(val)
            Scribe.Send(f"StyxScribeShared: Set: {i}{DELIM}{k}{DELIM}{v}")
    def __repr__(self):
        return repr(self._proxy)
    def __delitem__(self, key, sync=True):
        return self.__setitem__(key, NIL, sync)
    def __getitem__(self, key):
        try:
            return self._proxy[key]
        except KeyError:
            return NIL

class ProxyCall(Proxy):
    _repr = True
    def _marshall(self, obj):
        if callable(obj):
            self._proxy = obj
            SetName(self, obj.__name__)
        else:
            raise TypeError(f"{type(self)} cannot marshall type {type(obj)}")
    def _call(self, args):
        return self(*args)

@proxyType
@marshallType(dict, set)
class Table(ProxySet, dict):
    def _marshall(self, obj):
        if isinstance(obj, set):
            for k in obj:
                self[k] = True
        elif isinstance(obj, dict):
            for k,v in obj.items():
                self[k] = v
        else:
            raise TypeError(f"{type(self)} cannot marshall type {type(obj)}")
    def __setitem__(self, key, val, sync=True):
        key = marshall(key)
        if val is NIL:
            del self._proxy[key]
        else:
            val = marshall(val)
            self._proxy[key] = val
        if sync:
            self._shset(key, val)
    setdefault = notImplemented
    update = notImplemented
    pop = notImplemented
    popitem = notImplemented
    clear = notImplemented
    def __getattr__(self, name):
        if not name.startswith('_'):
            return self[name]
        return object.__getattribute__(self, name)
    def __setattr__(self, name, value):
        try:
            self.__getattribute__(name)
        except AttributeError:
            try:
                self.__getattr__(name)
            except AttributeError:
                pass
            else:
                self[name] = value
                return
        object.__setattr__(self, name, value)

@proxyType
@marshallType(list, tuple)
class Array(ProxySet, list):
    def _marshall(self, obj):
        enum = None
        try:
            enum = enumerate(obj)
        except TypeError:
            raise TypeError(f"{type(self)} cannot marshall type {type(obj)}")
        for k,v in enum:
            self[k] = v
    def __setitem__(self, key, val, sync=True):
        key = marshall(key)
        if val is NIL:
            d = key - len(self._proxy) + 1
            if d < 0:
                for i in range(d):
                    del self._proxy[-1]
            del self._proxy[key]
        else:
            if key > len(self._proxy):
                return
            val = marshall(val)
            if key == len(self._proxy):
                self._proxy.append(val)
            else:
                self._proxy[key] = val
        if sync:
            self._shset(key + 1, val)
    append = notImplemented
    extend = notImplemented
    insert = notImplemented
    remove = notImplemented
    sort = notImplemented
    reverse = notImplemented
    pop = notImplemented
    clear = notImplemented

@proxyType
class Args(Array):
    def __setitem__(self, key, val, sync=True):
        if val is NIL:
            if key >= len(self._proxy):
                return
            self._proxy[key] = NIL
        else:
            n = len(self._proxy)
            if key == 'n':
                val = marshall(val)
                if val >= n:
                    self._proxy.extend([NIL]*(val - n))
                if val < n:
                    self._proxy[::] = self._proxy[:val]
            else:
                key = marshall(key)
                val = marshall(val)
                if key > n:
                    self._proxy.extend([NIL]*(key - n)+[val])
                if key == len(self._proxy):
                    self._proxy.append(val)
                else:   
                    self._proxy[key] = val
        if sync:
            self._shset(key if key == 'n' else key + 1, val)
    def __getitem__(self, key):
        return self._proxy[key]

@proxyType
class KWArgs(Table):
    def __setitem__(self, key, val, sync=True):
        key = marshall(key)
        val = marshall(val)
        self._proxy[key] = val
        if sync:
            if isinstance(key, Number):
                key = key + 1
            self._shset(key, val)
    def __getitem__(self, key):
        return self._proxy[key]

@proxyType
@marshallType(_function)
class Action(ProxyCall, _function):
    def __call__(self, *args):
        if self._local:
            self._proxy(*args)
        else:
            i = lookup[self]
            a = Args(args)
            ai = lookup[a]
            Scribe.Send(f"StyxScribeShared: Act: {i}{DELIM}{ai}")

@proxyType
class KWAction(Action):
    def __call__(self, *args, **kwargs):
        if self._local:
            self._proxy(*args, **kwargs)
        else:
            i = lookup[self]
            a = dict(enumerate(args))
            a.update(kwargs)
            a = KWArgs(a)
            ai = lookup[a]
            Scribe.Send(f"StyxScribeShared: Act: {i}{DELIM}{ai}")
    def _call(self, args):
        kwargs = {k:v for k,v in args.items() if not isinstance(k,int) or k < 0}
        n = 1+max((-1,)+tuple(k for k in args.keys() if k not in kwargs))
        args = tuple(args[k] for k in range(n))
        return self(*args, **kwargs)

def marshaller(obj):
    if isinstance(obj,Proxy):
        return None
    for c,ts in marshallTypes.items():
        for t in ts:
            if isinstance(obj,t):
                return c
            if callable(obj) and t is _function:
                return c
    return None

def marshall(obj):
    m = marshaller(obj)
    if m is not None:
        return m(obj)
    if isinstance(obj,str):
        return obj.replace(DELIM,':')
    if isinstance(obj,float):
        if float.is_integer(obj):
            return int(obj)
    return obj

def encode(v):
    if isinstance(v, bool):
        return "!!" if v else "!"
    if isinstance(v, str):
        return "&" + v
    if isinstance(v, Number):
        return "#" + str(v)
    if isinstance(v, Proxy):
        return "@" + str(lookup[v])
    if v is NIL:
        return "*"
    raise TypeError(f"{v} of type {type(v)} cannot be encoded.")

def decode(s):
    t = s[:1]
    v = s[1:]
    if t == "&":
        return v
    if t == "#":
        if v.isdigit():
            return int(v)
        return float(v)
    if t == "@":
        return registry[-int(v)]
    if t == "!":
        return v[0] == '!'
    if t == "*":
        return NIL
    raise TypeError(s + " cannot be decoded.")

def handleNew(message):
    t, i = message.split(DELIM)
    val = proxyTypes[t](None, -int(i))
    return val

def handleSet(message):
    i, k, v = message.split(DELIM)
    try:
        s = registry[-int(i)]
    except KeyError:
        return
    k = decode(k)
    v = decode(v)
    s.__setitem__(k, v, False)

def handleDel(message):
    i = -int(message)
    obj = registry.get(i,None)
    if obj is not None:
        obj._alive = False
        try:
            del lookup[obj]
        except KeyError:
            pass
        if i != 0:
            try:
                del registry[i]
            except KeyError:
                pass

def handleAct(message):
    func, args = message.split(DELIM)
    func = registry[-int(func)]
    args = registry[-int(args)]
    func._call(args)

def handleName(message):
    obj, name = message.split(DELIM)
    obj = registry[-int(obj)]
    obj._name = decode(name)

def handleReset(message=None):
    global registry
    global lookup
    global Root
    if registry is not None:
        registry.clear()
    registry = {}
    if lookup is not None:
        lookup.clear()
    lookup = WeakKeyDictionary()
    Root = Table(None, 0)
    Scribe.Send("StyxScribeShared: Reset")

def SetName(proxy,name):
    proxy._name = name
    i = lookup[proxy]
    name = encode(name)
    Scribe.Send(f"StyxScribeShared: Name: {i}{DELIM}{name}")

def GetName(proxy):
    return proxy._name

def Load():
    Scribe.AddHook(handleReset, "StyxScribeShared: Reset", __name__)
    Scribe.AddHook(handleName, "StyxScribeShared: Name: ", __name__)
    Scribe.AddHook(handleNew, "StyxScribeShared: New: ", __name__)
    Scribe.AddHook(handleSet, "StyxScribeShared: Set: ", __name__)
    Scribe.AddHook(handleDel, "StyxScribeShared: Del: ", __name__)
    Scribe.AddHook(handleAct, "StyxScribeShared: Act: ", __name__)
    Scribe.IgnorePrefixes.append("StyxScribeShared: ")

File: Content/Mods/StyxScribeShared/StyxScribeShared.lua
ModUtil.Mod.Register( "StyxScribeShared" )

local delim = '¦'

local registry
local lookup
local objectData
local encode
local decode
local marshall

local classes = { }
local marshallTypes = { }
local marshallTypesOrder = { }
local NONE = { }
local ready = false

local function nop( ... ) return ... end

local function typeCall( m, f )
	local f = f or function( cls, ... )
		return cls:_new( ... )
	end
	local mm = getmetatable( m ) or { }
	mm.__call = f
	return setmetatable( m, mm )
end

local function marshallType( ... )
	local types = { ... }
	local m = table.remove( types )
	table.insert( marshallTypesOrder, m )
	for _, t in ipairs( types ) do
		mt = marshallTypes[ m ] or { }
		table.insert( mt, t )
		marshallTypes[ m ] = mt
	end
	return m
end

-- mirror the structure of the python classes as metatables (mostly)
local function class( name, ... )
	local metas = table.pack( ... )
	local n = metas.n
	local m = metas[ n ]
	for i = n - 1, 1, -1 do
		metas[ i + 1 ] = metas[ i ]
	end
	metas[ 1 ] = m
	local meta = { }
	for i = n, 1, -1 do
		for k, v in pairs( metas[ i ] ) do
			meta[ k ] = v
		end
	end
	if name ~= nil then
		meta._name = name
		classes[ name ] = meta
	end
	meta._parents = { table.unpack( metas, 2, n ) }
	return meta
end

local function new( m, ... )
	return m:_new( ... )
end

local _table = {
	_new = function( )
		return { }
	end
}

local _function = {
	_new = function( )
		return nop
	end
}

local Proxy = {
	_proxy = true,
	__gc = function( s )
		local data = objectData[ s ]
		if data ~= nil then
			data[ "alive" ] = false
		end
		local i = lookup[ s ]
		if i ~= nil then
			lookup[ s ] = nil
			if i ~= 0 then
				registry[ i ] = nil
				StyxScribe.Send( "StyxScribeShared: Del: " .. tostring( i ) )
			end
		end
		
	end,
	_new = function( m, ... )
		local s = setmetatable( { }, m )
		return (m._init or nop)( s, ... )
	end,
	_init = function( s, v, i )
		local meta = getmetatable( s )
		i = i or tonumber( ModUtil.ToString.Address( s ), 16 )
		objectData[ s ] = { }
		objectData[ s ][ "proxy" ] = meta:_newproxy( )
		objectData[ s ][ "alive" ] = true
		registry[ i ] = s
		lookup[ s ] = i
		objectData[ s ][ "root" ] = i == 0
		objectData[ s ][ "local" ] = i > 0
		if objectData[ s ][ "local" ] then
			StyxScribe.Send( "StyxScribeShared: New: " .. meta._name .. delim .. i )
		end
		if v then
			meta._marshall( s, v )
		end
		return s
	end,
	__tostring = function( s )
		local name = objectData[ s ][ "name" ]
		name = name and name .. ': ' or ""
		return name .. getmetatable( s )._name .. ': ' .. ModUtil.ToString.Address( s ) 
	end
}

local ProxySet = class( nil, Proxy, {
	_shset = function( s, k, v )
		local i = tostring( lookup[ s ] )
		k = encode( k )
		v = encode( v )
		StyxScribe.Send( "StyxScribeShared: Set: " .. i .. delim .. k .. delim .. v )
	end,
	_marshall = function( s, obj )
		for k, v in pairs( obj ) do
			s[ k ] = v
		end
	end,
	_newproxy = function( m )
		return { }
	end
} )

local ProxyCall = class( nil, Proxy, {
	_marshall = function( s, f )
		objectData[ s ][ "proxy" ] = f
	end,
	_newproxy = function( m )
		return nop
	end,
	_call = function( s, args )
		return s( table.unpack( args ) )
	end
} )

local Table = marshallType( "table", class( "Table", ProxySet, {
	__newindex = function( s, k, v, sync )
		k, v = marshall( k ), marshall( v )
		objectData[ s ][ "proxy" ][ k ] = v
		local meta = getmetatable( s )
		if sync == nil or sync then
			meta._shset( s, k, v )
		end
	end,
	__index = function( s, k )
		return objectData[ s ][ "proxy" ][ k ]
	end,
	__len = function( s )
		return #objectData[ s ][ "proxy" ]
	end,
	__next = function( s, ... )
		return next( objectData[ s ][ "proxy" ], ... )
	end,
	__inext = function( s, ... )
		return inext( objectData[ s ][ "proxy" ], ... )
	end,
	__pairs = function( s, ... )
		return pairs( objectData[ s ][ "proxy" ], ... )
	end,
	__ipairs = function( s, ... )
		return ipairs( objectData[ s ][ "proxy" ], ... )
	end
} ) )

local Array = marshallType( "table", typeCall( class( "Array", ProxySet, {
	__newindex = function( s, k, v, sync )
		k, v = marshall( k ), marshall( v )
		local proxy = objectData[ s ][ "proxy" ]
		local n = #proxy
		if type( k ) ~= "number" or math.floor( k ) ~= k then
			error( "Array index must be an integer" , 2 )
		end
		if k > n + 1 or k < 1 then
			error( "Array index " .. tostring( k ) .." out of bounds" , 2 )
		end
		proxy[ k ] = v
		local meta = getmetatable( s )
		if sync == nil or sync then
			meta._shset( s, k - 1, v )
		end
	end,
	__index = function( s, k )
		return objectData[ s ][ "proxy" ][ k ]
	end,
	__len = function( s )
		return #objectData[ s ][ "proxy" ]
	end,
	__next = function( s, ... )
		return inext( objectData[ s ][ "proxy" ], ... )
	end,
	__inext = function( s, ... )
		return inext( objectData[ s ][ "proxy" ], ... )
	end,
	__pairs = function( s, ... )
		return ipairs( objectData[ s ][ "proxy" ], ... )
	end,
	__ipairs = function( s, ... )
		return ipairs( objectData[ s ][ "proxy" ], ... )
	end
} ) ) )

local Args = marshallType( "table", typeCall( class( "Args", ProxySet, {
	__newindex = function( s, k, v, sync )
		k, v = marshall( k ), marshall( v )
		local proxy = objectData[ s ][ "proxy" ]
		local n = proxy.n or #proxy
		if k == 'n' then
			if v < n then
				for i = v + 1, n, 1 do
					proxy[ i ] = nil
				end
			end
		else
			if type( k ) ~= "number" or math.floor( k ) ~= k then
				error( "Args index must be an integer" , 2 )
			end
			if k < 1 then
				error( "Args index " .. tostring( k ) .." out of bounds" , 2 )
			end
			if k >= n then
				proxy.n = k
			end
		end
		proxy[ k ] = v
		local meta = getmetatable( s )
		if sync == nil or sync then
			meta._shset( s, k ~= 'n' and k - 1 or k, v )
		end
	end,
	__index = function( s, k )
		return objectData[ s ][ "proxy" ][ k ]
	end,
	__len = function( s )
		return objectData[ s ][ "proxy" ][ 'n' ]
	end,
	__next = function( s, ... )
		return next( objectData[ s ][ "proxy" ], ... )
	end,
	__inext = function( s, ... )
		local k, v = next( objectData[ s ][ "proxy" ], ... )
		if k == 'n' then
			k, v = next( objectData[ s ][ "proxy" ], k )
		end
		return k, v
	end,
	__pairs = function( s, ... )
		return pairs( objectData[ s ][ "proxy" ], ... )
	end,
	__ipairs = function( s, ... )
		return qrawipairs( objectData[ s ][ "proxy" ], ... )
	end
} ), function( cls, ... ) return cls:_new( ... ) end ) )

local KWArgs = marshallType( "table", typeCall( class( "KWArgs", Table, {
	__newindex = function( s, k, v, sync )
		k, v = marshall( k ), marshall( v )
		objectData[ s ][ "proxy" ][ k ] = v
		local meta = getmetatable( s )
		if sync == nil or sync then
			if type( k ) == "number" then
				k = k - 1
			end
			meta._shset( s, k, v )
		end
	end
} ) ) )

local Action = marshallType( "function", typeCall( class( "Action", ProxyCall, {
	__call = function( s, ... )
		if objectData[ s ][ "local" ] then
			objectData[ s ][ "proxy" ]( ... )
		else
            local i = tostring( lookup[ s ] )
            local a = new( Args, table.pack( ... ) )
            local ai = tostring( lookup[ a ] )
            StyxScribe.Send( "StyxScribeShared: Act: " .. i .. delim .. ai )
		end
	end
} ) ) )

local KWAction = typeCall( class( "KWAction", Action, { 
	__call = function( s, kwargs )
		if objectData[ s ][ "local" ] then
			objectData[ s ][ "proxy" ]( kwargs )
		else
            local i = tostring( lookup[ s ] )
            local a = new( KWArgs, kwargs )
            local ai = tostring( lookup[ a ] )
            StyxScribe.Send( "StyxScribeShared: Act: " .. i .. delim .. ai )
		end
	end,
	_call = function( s, args )
		return s( args )
	end
} ) )

--https://stackoverflow.com/a/33312901
local function split( text, delim )
    -- returns an array of fields based on text and delimiter (one character only)
    local result = { }
    local magic = "().%+-*?[]^$"

    if delim == nil then
        delim = "%s"
    elseif string.find( delim, magic, 1, true ) then
        -- escape magic
        delim = "%" .. delim
    end

    local pattern = "[^" .. delim .. "]+"
    for w in string.gmatch( text, pattern ) do
        table.insert( result, w )
    end
    return result
end

local function marshaller( obj )
	local t = type( obj )
	local m = getmetatable( obj )
    if m and m._proxy then return end
    for _, m in ipairs( marshallTypesOrder ) do
		for _, _t in pairs( marshallTypes[ m ] ) do
			if t == _t then
				return m
			end
		end
	end
end

function marshall( obj )
    if obj == NONE then return nil end
    local m = marshaller( obj )
    if m then return new( m, obj ) end
    if type( obj ) == "string" then
        return obj:gsub( delim, ':' )
	end
    return obj
end

function decode( s )
	local t, v = s:sub(1,1), s:sub(2)
	if t == "&" then return v end
	if t == "#" then return tonumber( v ) end
	if t == "@" then return registry[ -tonumber( v ) ] end
	if t == "!" then return v == "!" end
	if t == "*" then return nil end
	error( s .. " cannot be decoded.", 2 )
end

function encode( v )
	local t, m = type( v ), getmetatable( v )
	if t == "string" then return "&" .. v end
	if t == "number" then return "#" .. tostring( v ) end
	if m and m._proxy then return "@" .. tostring( lookup[ v ] ) end
	if t == "boolean" then return v and "!!" or "!" end
	if v == nil or v == NONE then return "*" end
	error( tostring( v ) .. " cannot be encoded.", 2 )
end

local function handleNew( message )
	if not ready then return end
	local name, id = table.unpack( split( message, delim ) )
	return new( classes[ name ], nil, -tonumber( id ) )
end

local function handleDel( message )
	if not ready then return end
	id = -tonumber( message )
	obj = registry[ id ]
	if obj ~= nil then
		objectData[ obj ][ "alive" ] = false
		registry[ id ] = nil
	end
end

local function handleSet( message )
	if not ready then return end
	local id, key, value = table.unpack( split( message, delim ) )
	local obj = registry[ -tonumber( id ) ]
	key = decode( key )
	value = decode( value )
	local meta = getmetatable( obj )
	meta.__newindex( obj, key, value, false )
end

local function handleAct( message )
	if not ready then return end
	local func, args = table.unpack( split( message, delim ) )
    func = registry[ -tonumber( func ) ]
    args = registry[ -tonumber( args ) ]
    getmetatable(func)._call( func, args )
end

local function handleName( message )
	if not ready then return end
	local id, name = table.unpack( split( message, delim ) )
	local obj = registry[ -tonumber( id ) ]
	objectData[ obj ][ "name" ] = decode( name )
end

local function handlePyReset( )
	ready = true
end

local function handleLuaReset( )
	ready = false
	registry = { }
	lookup = setmetatable( { }, { __mode = "k" } )
	objectData = setmetatable( { }, { __mode = "k" } )
	StyxScribeShared.Root = new( Table, nil, 0 )
	StyxScribe.Send( "StyxScribeShared: Reset" )
end

function StyxScribeShared.SetName( proxy, name )
	objectData[ proxy ][ "name" ] = name
	local id = tostring( lookup[ proxy ] )
	StyxScribe.Send( "StyxScribeShared: Name: " .. id .. delim .. encode( name ) )
end

function StyxScribeShared.GetName( proxy )
	return objectData[ obj ][ "name" ]
end

StyxScribeShared.Internal = ModUtil.UpValues( function( )
	return registry, lookup, delim, objectData, split, class, new, nop,
		marshallType, marshallTypes, marshaller, marshall, _table, _function,
		Proxy, ProxySet, ProxyCall, typeCall, decode, encode, ready,
		handleNew, handleSet, handleReset, handleAct, handleDel, handlePyReset, handleName,
		NONE, Table, Array, Args, Action, KWArgs, KWAction
end )

ModUtil.Table.Merge( StyxScribeShared, {
	NONE = NONE, Table = Table, Array = Array, Action = Action, KWArgs = KWArgs, KWAction = KWAction
} )

StyxScribe.AddHook( handlePyReset, "StyxScribeShared: Reset", StyxScribeShared )
StyxScribe.AddHook( handleName, "StyxScribeShared: Name: ", StyxScribeShared )
StyxScribe.AddHook( handleNew, "StyxScribeShared: New: ", StyxScribeShared )
StyxScribe.AddHook( handleSet, "StyxScribeShared: Set: ", StyxScribeShared )
StyxScribe.AddHook( handleDel, "StyxScribeShared: Del: ", StyxScribeShared )
StyxScribe.AddHook( handleAct, "StyxScribeShared: Act: ", StyxScribeShared )
handleLuaReset( )